pkg/controllers/clusterca/clusterca_controller.go

```diff
@@ -0,0 +1,122 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ClusterCaReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func SetupWithManager(mgr manager.Manager) error {
+	if err := add(mgr, newReconciler(mgr)); err != nil {
+		klog.Errorf("Failed to create clusterrole controller, %v", err)
+		return err
+	}
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ClusterCaReconciler{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterca-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &clusterinfov1beta1.ManagedClusterInfo{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ClusterCaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clusterinfo := &clusterinfov1beta1.ManagedClusterInfo{}
+
+	err := r.client.Get(ctx, req.NamespacedName, clusterinfo)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	cluster := &clusterv1.ManagedCluster{}
+
+	err = r.client.Get(ctx, types.NamespacedName{Name: clusterinfo.Name}, cluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	updatedClientConfig, needUpdate := updateClientConfig(cluster.Spec.ManagedClusterClientConfigs, clusterinfo.Status.DistributionInfo.OCP.ManagedClusterClientConfig)
+	if !needUpdate {
+		return ctrl.Result{}, nil
+	}
+
+	cluster.Spec.ManagedClusterClientConfigs = updatedClientConfig
+	err = r.client.Update(ctx, cluster, &client.UpdateOptions{})
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+//updateClientConfig merge config from clusterinfoconfigs to clusterconfigs
+func updateClientConfig(clusterConfigs []clusterv1.ClientConfig, clusterinfoConfig clusterinfov1beta1.ClientConfig) ([]clusterv1.ClientConfig, bool) {
+	var returnConfig = clusterConfigs
+	needUpdate := false
+	exist := false
+
+	if reflect.DeepEqual(clusterinfoConfig, clusterinfov1beta1.ClientConfig{}) {
+		return returnConfig, false
+	}
+
+	for i, cluConfig := range clusterConfigs {
+		if cluConfig.URL != clusterinfoConfig.URL {
+			continue
+		}
+		if !reflect.DeepEqual(cluConfig.CABundle, clusterinfoConfig.CABundle) {
+			needUpdate = true
+			returnConfig[i].CABundle = clusterinfoConfig.CABundle
+		}
+		exist = true
+	}
+
+	//already have ca bundle in managedcluster config
+	if exist {
+		return returnConfig, needUpdate
```

Comment:
what does this section mean?

Reply:
> what does this section mean?
it checked if their already have a ca in managedcluster's config, and if the existing ca need updated

The "returnConfig" stored all managedcluster's config, and even updated if need, "needUpdate" check if it has been updated.